### ISSUE TITLE
Negative "Loss" values for Example 8

### ISSUE THREAD

**Issue by jatinchowdhury18** [gdalsanto/flamo#13]
Hello, thanks for the wonderful library!

I was running the `examples/e8_colorless_fdn.py` script with the "Scattering Matrix" code un-commented, and I ran into an issue, where the training loss became negative. Here's the output of the script:

```
❯ python flamo/examples/e8_colorless_fdn.py --max_epochs 100
cuda not available, will use cpu
Training:   0%|                                                                                              | 0/100 [00:00<?, ?it/s]epoch:   0 - train_loss: 7.5704 - test_loss: 4.6798 - time: 32.9247 s
Training:   1%|▊                                                                                     | 1/100 [00:32<54:19, 32.93s/it]epoch:   1 - train_loss: 2.6696 - test_loss: 1.6901 - time: 35.0966 s
Training:   2%|█▋                                                                                    | 2/100 [01:08<55:51, 34.20s/it]epoch:   2 - train_loss: 0.8650 - test_loss: 0.5121 - time: 35.5264 s
Training:   3%|██▌                                                                                   | 3/100 [01:43<56:16, 34.81s/it]epoch:   3 - train_loss: 0.1501 - test_loss: 0.0280 - time: 34.5465 s
Training:   4%|███▍                                                                                  | 4/100 [02:18<55:31, 34.70s/it]epoch:   4 - train_loss: -0.1405 - test_loss: -0.1736 - time: 34.8565 s
Training:   5%|████▎                                                                                 | 5/100 [02:52<55:02, 34.76s/it]epoch:   5 - train_loss: -0.2578 - test_loss: -0.2571 - time: 33.3632 s
Training:   6%|█████▏                                                                                | 6/100 [03:26<53:42, 34.29s/it]epoch:   6 - train_loss: -0.3041 - test_loss: -0.2916 - time: 32.3397 s
Training:   7%|██████                                                                                | 7/100 [03:58<52:09, 33.65s/it]epoch:   7 - train_loss: -0.3226 - test_loss: -0.3069 - time: 32.4975 s
Training:   8%|██████▉                                                                               | 8/100 [04:31<51:02, 33.28s/it]epoch:   8 - train_loss: -0.3313 - test_loss: -0.3153 - time: 32.4883 s
Training:   9%|███████▋                                                                              | 9/100 [05:03<50:06, 33.03s/it]epoch:   9 - train_loss: -0.3373 - test_loss: -0.3221 - time: 34.4814 s
Training:  10%|████████▌                                                                            | 10/100 [05:38<50:13, 33.48s/it]epoch:  10 - train_loss: -0.3434 - test_loss: -0.3290 - time: 34.3016 s
```
Is this to be expected? In most training/optimization systems that I've worked with, the loss functions are always expected to generate a positive output.

For the moment, I've enforced this in a pretty crude way by modifying the trainer to take the absolute value of each "criterion", e.g.
```diff
-temp = criterion(estimations, targets, self.net)
+temp = torch.abs(criterion(estimations, targets, self.net))
self.train_loss_log[criterion.__class__.__name__].append(temp.item())
loss += alpha*temp
```

**Comment by gdalsanto** [comment]
Hi Jatin, 
thanks for opening the issue!

The negative loss value is most probably due to the [sparsity loss](https://github.com/gdalsanto/flamo/blob/2b7181a54f192817bf650ee8590e16633aba1f62/flamo/optimize/loss.py#L8) which was designed for the scalar feedback matrix and does not account for the filter length as an additional dimension. During our tests, we found that a negative loss value did not pose an issue for optimization. However, as it can cause confusion and is suboptimal for training, we have [updated](https://github.com/gdalsanto/flamo/blob/c23668ce9dff30318ac8bac92f051476d45201c4/flamo/optimize/loss.py#L17) the sparsity loss to compute the mean over the filter length dimension.
This should create only positive loss values.

**Comment by jatinchowdhury18** [comment]
Sounds good! Thanks for the explanation.
